src/Data/Morpheus/Types/Internal/Subscription.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE NamedFieldPuns          #-}
+{-# LANGUAGE FlexibleContexts        #-}
+{-# LANGUAGE FlexibleInstances       #-}
+{-# LANGUAGE UndecidableInstances    #-}
+{-# LANGUAGE MultiParamTypeClasses   #-}
+{-# LANGUAGE DataKinds               #-}
+{-# LANGUAGE GADTs                   #-}
+{-# LANGUAGE TypeFamilies            #-}
+{-# LANGUAGE ScopedTypeVariables     #-}
+
+module Data.Morpheus.Types.Internal.Subscription
+  ( Client
+  , connect
+  , disconnect
+  , toOutStream
+  , runStreamWS
+  , runStreamHTTP
+  , Stream
+  , Scope(..)
+  , Input(..)
+  , API(..)
+  , acceptApolloRequest
+  , publish
+  , Store(..)
+  , initDefaultStore
+  )
+where
+
+import           Control.Concurrent             ( readMVar
+                                                , newMVar
+                                                , modifyMVar_
+                                                )
+import           Data.UUID.V4                   ( nextRandom )
+
+-- MORPHEUS
+import           Data.Morpheus.Types.Internal.Operation
+                                                (empty)
+import           Control.Monad.IO.Class         ( MonadIO(..) )
+import           Data.Morpheus.Types.Internal.Resolving
+                                                ( GQLChannel(..) )
+import           Data.Morpheus.Types.Internal.Subscription.Apollo
+                                                ( acceptApolloRequest )
+import           Data.Morpheus.Types.Internal.Subscription.ClientStore
+                                                ( delete 
+                                                , Client
+                                                , publish
+                                                , ClientStore
+                                                )
+import           Data.Morpheus.Types.Internal.Subscription.Stream
+                                                ( toOutStream
+                                                , runStreamWS
+                                                , runStreamHTTP
+                                                , Stream
+                                                , Scope(..)
+                                                , Input(..)
+                                                , API(..)
+                                                )
+ 
+connect :: IO (Input 'Ws)
+connect = Init <$> nextRandom
+
+disconnect:: Scope 'Ws e m -> Input 'Ws -> m ()
+disconnect WS { update }  (Init clientID)  = update (delete clientID)
+
+
+-- | shared GraphQL state between __websocket__ and __http__ server,
+-- stores information about subscriptions
+data Store e m = Store 
+  { readStore :: m (ClientStore e m)
+  , writeStore :: (ClientStore e m -> ClientStore e m) -> m ()
+  }
```

Comment:
So if, say, my PubSub store is RabbitMQ or Redis, this is the interface it has to respect to be able to replace the previous inmemory store ?

Reply:
but if you want use your own monad. for example to log output. or you don't want to use default 
implementation with `MVar`  you can implement Store where you provide `write` and `read` .

Reply:
So with the current implementation I cannot use an external pubsub service ? (that's what I'm after, but I don't understand everything in the current subscription design).
Say I have a subscribe function in haskell which listen to an external data store. I'd like to be able to do something like :
```haskell

subscription :: IO ()
subscription = do
  subscribeExternalService callback
  forever $ threadDelay maxBound  

callback :: SomeDataType -> IO ()
callback someDataToSendToClient = sendToClient someDataToSendToClient

```

Do you think this is possible with the current design ?

Reply:
the store only remembers local server  connections( with subscriptions). there is no point to use any external store for it. 

if you want synchronize it with external store. you can forward received event from external source to current server and trigger it with  
`publishEventWith`.

Reply:
That makes sense, thanks !

Reply:
@theobat 
finally i decided to define simple websocketApp

```hs
(wsApp, publish) <- webSocketsApp api
-- that can be used in 
httpAppWithEffect publish api
```
less code and less complexity for user.

what do you think?

Reply:
Yeah I think it's great, I still have to wrap my head around all the subscription stuff it's not entirely clear to me yet. I'll try to add haddock comments to the code cause to document my digging (I'll start with this PR merged).